src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp

```diff
@@ -0,0 +1,261 @@
+    // The current thread should block the subsequent background thread until GC completes
+    // If the call to EnumerateGCHeapObjects doesn't wait for GC to complete, it should
+    // observe IsGCStartSleeping() as true during the heap walk.
+    _threadList.emplace_back(std::thread([instance]()
+                {
+                    printf("EnumerateGCHeapObject on native background thread\n");
+                    instance->ValidateEnumerateGCHeapObjects(S_OK);
+                }));
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+    _gcStartSleeping = FALSE;
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionFinished()
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionFinished\n");
+    return S_OK;
+}
+
+bool GCHeapEnumerationProfiler::IsGCStartSleeping()
+{
+    return _gcStartSleeping;
+}
+
+HRESULT GCHeapEnumerationProfiler::Shutdown()
+{
+    // make sure async operations have finished
+    for (auto& t : _threadList) {
+        t.join();
+    }
+    Profiler::Shutdown();
+
+    if (_expectedExceptions == 1)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: PASS: Encountered exception as expected.\n");
+        printf("PROFILER TEST PASSES\n");
+        fflush(stdout);
+        return S_OK;
+    }
+
+    if (_objectsCount < 100)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: FAIL: Expected at least 100 objects, got %d\n", _objectsCount.load());
+        IncrementFailures();
+    }
+
+    if (_customGCHeapObjectTypesCount != 1)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: FAIL: Expected 1 custom GCHeapObject type, got %d\n", _customGCHeapObjectTypesCount.load());
+        IncrementFailures();
+    }
+
+    if (_failures == 0)
+    {
+        printf("PROFILER TEST PASSES\n");
+    }
+    else
+    {
+        // failures were printed earlier when _failures was incremented
+    }
+    fflush(stdout);
+
+    return S_OK;
+}
+
+struct CallbackState
+{
+    std::atomic<int>* objectsCount;
+    GCHeapEnumerationProfiler *instance;
+    std::atomic<int>* customGcHeapObjectTypeCount;
+};
+
+static BOOL STDMETHODCALLTYPE heap_walk_fn(ObjectID object, void* callbackState)
+{
+    CallbackState* state = static_cast<CallbackState*>(callbackState);
+
+    if (state->instance->IsGCStartSleeping())
+    {
+        printf("Error: no callbacks expected during GC\n");
+        state->instance->IncrementFailures();
+        return FALSE;
+    }
+
+    state->objectsCount->fetch_add(1, std::memory_order_relaxed);
+
+    ClassID classId{0};
+    HRESULT hr = state->instance->pCorProfilerInfo->GetClassFromObject(object, &classId);
+    if (FAILED(hr))
+    {
+        printf("Error: failed to get class ID from object with ID 0x%p. hr=0x%x\n", (void *)object, hr);
+        state->instance->IncrementFailures();
+        return FALSE;
+    }
+
```

Comment:
Can we also validate that we can get the class layout here (using GetClassLayout) since ultimately we'll want to read fields on these types?

Reply:
I'm able to call that API as well from the heap walk fn, and for a majority of the classId's I'm hitting these Invalid args https://github.com/dotnet/runtime/blob/041375910885f3a30ccf0bb8b751566dae7043f4/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L7848-L7860. Also not quite sure yet how to extract the information from `COR_FIELD_OFFSET` to get useful field info

Reply:
These layout APIs expect you to divide up heap objects into strings, arrays, and everything else. GetClassLayout() only returns a successful result on the 'everything else' category. To determine if a type is an array you can call ICorProfilerInfo::IsArrayClass(). To identify string you can check if the type name is "System.String". A more efficient profiler (ie not test code) would probably want to cache the unique ClassID value for string and then do a direct comparison on the ClassID to identify strings.

Reply:
> Also not quite sure yet how to extract the information from COR_FIELD_OFFSET to get useful field info

COR_FIELD_OFFSET has a FieldDef metadata token you can look up with metadata APIs to get a name. It also has an offset which represents the memory offset from the start of the object where data for that field is stored. I think its sufficient just to verify that GetClassLayout returned S_OK and assume the layout information it is giving you is correct.

Reply:
Added validation to check that class field information can be extracted from all non-string, non-array objects found on the gc heap